gr-digital/lib/ofdm_cyclic_prefixer_impl.cc

```diff
@@ -31,135 +31,147 @@ namespace gr {
+        d_delay_line(0, 0),
+        d_len_tag_key(len_tag_key)
     {
-      set_relative_rate(d_output_size);
-
-      // Flank of length 1 would just be rectangular
-      if (d_rolloff_len == 1) {
-	d_rolloff_len = 0;
+      for(const int& cp_len : d_cp_lengths) {
+        d_cp_max = std::max(d_cp_max, cp_len);
       }
-      if (d_rolloff_len) {
-	d_delay_line.resize(d_rolloff_len-1, 0);
-	if (rolloff_len > d_cp_size) {
-	  throw std::invalid_argument("cyclic prefixer: rolloff len must smaller than the cyclic prefix.");
-	}
-	// The actual flanks are one sample shorter than d_rolloff_len, because the
-	// first sample of the up- and down flank is always zero and one, respectively
-	for (int i = 1; i < d_rolloff_len; i++) {
-	  d_up_flank[i-1]   = 0.5 * (1 + cos(M_PI * i/rolloff_len - M_PI));
-	  d_down_flank[i-1] = 0.5 * (1 + cos(M_PI * (rolloff_len-i)/rolloff_len - M_PI));
-	}
+      // Sanity
+      for(unsigned i=0; i<d_cp_lengths.size(); i++) {
+        if (d_cp_lengths[i] != 0) {
+          break;
+        }
+        if (i == d_cp_lengths.size()-1) {
+          throw std::invalid_argument(this->alias() + std::string(": It does not make sense to only have one or a couple of CPs with length 0."));
+        }
       }
-
-      if (len_tag_key.empty()) {
-	set_output_multiple(d_output_size);
-      } else {
-	set_tag_propagation_policy(TPP_DONT);
+      if (d_cp_lengths.empty()) {
+        throw std::invalid_argument(this->alias() + std::string(": CP lengths vector can not be empty. Please enter at least (d,) (a single Python tuple) with d being a integer CP length."));
+      }
+      // Give the buffer allcoator and scheduler a hint about the ratio between input and output.
+      set_relative_rate(d_cp_max + d_fft_len);
+      // Flank of length 1 would just be rectangular.
+      if(d_rolloff_len == 1) {
+        throw std::invalid_argument(this->alias() + std::string(": A rolloff length of 1 does not make sense, as it results in a boxcar function."));
+      }
+      if(d_rolloff_len) {
+        d_delay_line.resize(d_rolloff_len-1, 0);
+        // More sanity
+        for(const int& cp_len : d_cp_lengths) {
+          if(d_rolloff_len > cp_len) {
+            throw std::invalid_argument(this->alias() + std::string(": Rolloff length must be smaller than the cyclic prefix."));
+          }
+        }
+        /* The actual flanks are one sample shorter than d_rolloff_len, because the
+           first sample of the up- and down flank is always zero and one, respectively.*/
+        for(int i = 1; i < d_rolloff_len; i++) {
+          d_up_flank[i-1]   = 0.5 * (1 + cos(M_PI * i/rolloff_len - M_PI));
+          d_down_flank[i-1] = 0.5 * (1 + cos(M_PI * (rolloff_len-i)/rolloff_len - M_PI));
+        }
+      }
+      if(d_len_tag_key.empty()) {
+        // noutput_items is set to be a multiple of the largest possible output size.
+        // It is always OK to return less (in case of the shorter CP).
+        set_output_multiple(d_cp_max);
```

Comment:
If you have d_cp_max, you might as well also collect d_cp_min for some of the sanity checks above. Saves you a loop.